### PR TITLE
Deny hashing ty/re/ct inference variables

### PR DESCRIPTION
**Pull request by compiler-errors**
cc @cjgillot and https://github.com/rust-lang/rust/pull/102695#issuecomment-1275706528
r? @lcnr

best reviewed one commit at a time, mostly because the second commit that fixes `ClosureOutlivesRequirement` is mostly noise because of losing its `<'tcx>` lifetime parameter.

### CODE REVIEW

File: compiler/rustc_middle/src/ty/context.rs [hunk 1]
```diff
@@ -198,9 +198,11 @@ impl<'tcx> CtxtInterners<'tcx> {
                 .intern(kind, |kind| {
                     let flags = super::flags::FlagComputation::for_kind(&kind);
 
-                    // It's impossible to hash inference regions (and will ICE), so we don't need to try to cache them.
+                    // It's impossible to hash inference variables (and will ICE), so we don't need to try to cache them.
                     // Without incremental, we rarely stable-hash types, so let's not do it proactively.
-                    let stable_hash = if flags.flags.intersects(TypeFlags::HAS_RE_INFER)
+                    let stable_hash = if flags
+                        .flags
+                        .intersects(TypeFlags::HAS_RE_INFER | TypeFlags::HAS_TY_INFER)
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/context.rs:205]:
```suggestion
.intersects(TypeFlags::HAS_RE_INFER | TypeFlags::HAS_TY_INFER | TypeFlags::HAS_CT_INFER)
```

**Review by lcnr — COMMENTED** [review]
the same should hold for const infer vars

**Comment by lcnr** [comment]
once we also forbid const infer vars

r=me

**Comment by compiler-errors** [comment]
I have no idea why this would introduce a regression :thinking:

**Comment by nnethercote** [comment]
The perf changes are small and the wins balance out the losses. No need to stress on this one, I think.

@rustbot label: +perf-regression-triaged

**Pull request merged**